libp2p-kad/src/kbucket.rs

```diff
@@ -0,0 +1,474 @@
+use std::mem;
+use std::slice::Iter as SliceIter;
+use std::time::{Duration, Instant};
+use std::vec::IntoIter as VecIntoIter;
+
+/// Maximum number of nodes in a bucket.
+pub const MAX_NODES_PER_BUCKET: usize = 20;
+
+/// Table of k-buckets with interior mutability.
+#[derive(Debug)]
+pub struct KBucketsTable<Id, Val> {
+    my_id: Id,
+    tables: Vec<Mutex<KBucket<Id, Val>>>,
+    // The timeout when pinging the first node after which we consider that it no longer responds.
+    ping_timeout: Duration,
+}
+
+impl<Id, Val> Clone for KBucketsTable<Id, Val>
+where
+    Id: Clone,
+    Val: Clone,
+{
+    #[inline]
+    fn clone(&self) -> Self {
+        KBucketsTable {
+            my_id: self.my_id.clone(),
+            tables: self.tables
+                .iter()
+                .map(|t| t.lock().clone())
+                .map(Mutex::new)
+                .collect(),
+            ping_timeout: self.ping_timeout.clone(),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+struct KBucket<Id, Val> {
+    // Nodes are always ordered from oldest to newest.
+    // Note that we will very often move elements to the end of this. No benchmarking has been
+    // performed, but it is very likely that a `SmallVec` is the most performant data structure.
+    nodes: SmallVec<[Node<Id, Val>; MAX_NODES_PER_BUCKET]>,
+
+    // Node received when the bucket was full. Will be added to the list if the first node doesn't
+    // respond in time to our ping. The second element is the time when the pending node was added.
+    // If it is too much in the past, then we drop the first node and add the pending node to the
+    // end of the list.
+    pending_node: Option<(Node<Id, Val>, Instant)>,
+
+    // Last time this bucket was updated.
+    last_update: Instant,
+}
+
+impl<Id, Val> KBucket<Id, Val> {
+    // Puts the kbucket into a coherent state.
+    // If a node is pending and the timeout has expired, removes the first element of `nodes`
+    // and pushes back the node in `pending_node`.
+    fn flush(&mut self, timeout: Duration) {
+        if let Some((_, instant)) = self.pending_node {
+            if instant.elapsed() >= timeout {
+                let (pending_node, _) = self.pending_node.take().unwrap();
+                let _ = self.nodes.remove(0);
+                self.nodes.push(pending_node);
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+struct Node<Id, Val> {
+    id: Id,
+    value: Val,
+}
+
+/// Trait that must be implemented on types that can be used as an identifier in a k-bucket.
+pub trait KBucketsPeerId: Eq + Clone {
+    /// Distance between two peer IDs.
+    type Distance: Ord;
+
+    /// Computes the XOR of this value and another one.
+    fn distance_with(&self, other: &Self) -> Self::Distance;
+
+    /// Returns then number of bits that are necessary to store the distance between peer IDs.
+    /// Used for pre-allocations.
+    ///
+    /// > **Note**: Please do not return 0. That would be a bad idea.
```

Comment:
Just say what happens if you return zero. Ideally, panic.

Also, why can't this be `mem::size_of::<<T as KBucketsPeerId>::Distance>() * 8`? Do we need sub-byte precision?

Reply:
It's likely that the `Distance` becomes an `enum` or something similar in the future, in which case we don't want to use it directly.